precompiles/gov/gov.go

```diff
@@ -0,0 +1,144 @@
+// Copyright Tharsis Labs Ltd.(Evmos)
+// SPDX-License-Identifier:ENCL-1.0(https://github.com/evmos/evmos/blob/main/LICENSE)
+
+package gov
+
+import (
+	"embed"
+
+	"cosmossdk.io/log"
+	storetypes "cosmossdk.io/store/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	authzkeeper "github.com/cosmos/cosmos-sdk/x/authz/keeper"
+	govkeeper "github.com/cosmos/cosmos-sdk/x/gov/keeper"
+
+	"github.com/ethereum/go-ethereum/accounts/abi"
+	"github.com/ethereum/go-ethereum/common"
+
+	cmn "github.com/evmos/evmos/v20/precompiles/common"
+	"github.com/evmos/evmos/v20/x/evm/core/vm"
+	evmtypes "github.com/evmos/evmos/v20/x/evm/types"
+)
+
+var _ vm.PrecompiledContract = &Precompile{}
+
+// Embed abi json file to the executable binary. Needed when importing as dependency.
+//
+//go:embed abi.json
+var f embed.FS
+
+// Precompile defines the precompiled contract for gov.
+type Precompile struct {
+	cmn.Precompile
+	govKeeper govkeeper.Keeper
+}
+
+// LoadABI loads the gov ABI from the embedded abi.json file
+// for the gov precompile.
+func LoadABI() (abi.ABI, error) {
+	return cmn.LoadABI(f, "abi.json")
+}
+
+// NewPrecompile creates a new gov Precompile instance as a
+// PrecompiledContract interface.
+func NewPrecompile(
+	govKeeper govkeeper.Keeper,
+	authzKeeper authzkeeper.Keeper,
+) (*Precompile, error) {
+	abi, err := LoadABI()
+	if err != nil {
+		return nil, err
+	}
+
+	p := &Precompile{
+		Precompile: cmn.Precompile{
+			ABI:                  abi,
+			AuthzKeeper:          authzKeeper,
+			KvGasConfig:          storetypes.KVGasConfig(),
+			TransientKVGasConfig: storetypes.TransientGasConfig(),
+			ApprovalExpiration:   cmn.DefaultExpirationDuration, // should be configurable in the future.
+		},
+		govKeeper: govKeeper,
+	}
+
+	// SetAddress defines the address of the gov precompiled contract.
+	p.SetAddress(common.HexToAddress(evmtypes.GovPrecompileAddress))
+
+	return p, nil
+}
+
+// RequiredGas calculates the precompiled contract's base gas rate.
+func (p Precompile) RequiredGas(input []byte) uint64 {
+	// NOTE: This check avoid panicking when trying to decode the method ID
+	if len(input) < 4 {
+		return 0
+	}
```

Comment:
Not specific to this precompile, but why are we checking this here instead of here:https://github.com/evmos/evmos/blob/e95e3b683435f51ae80ae80242a2f3f5ecbc6a8d/x/evm/core/vm/contracts.go#L288-L288?

Reply:
previously this was on the geth fork, and I guess we didn't consider placing it there, or we encountered this issue on an edge case, and applied that fix. Now that we have the `core/vm` imported into the codebase, makes sense to add the check once there.

Reply:
let's start defining a constant value for the gas for each query and tx moving forward